sdk/examples/xo_python/sawtooth_xo/processor/handler.py

```diff
@@ -214,66 +217,157 @@ def _store_game_data(state_store, name, board, state, player_1, player_2):
+    x_wins = _is_win(board, MARK_1)
+    o_wins = _is_win(board, MARK_2)
+
+    if x_wins and o_wins:
+        raise InternalError(
+            'Two winners (there can be only one)')
+
+    elif x_wins:
+        return P1_WIN
+
+    elif o_wins:
+        return P2_WIN
+
+    elif EMPTY_SPACE not in board:
+        return TIE
+
+    elif state == P1_NEXT:
+        return P2_NEXT
+
+    elif state == P2_NEXT:
+        return P1_NEXT
+
+    elif state in END_STATES:
+        return state
+
+    else:
+        raise InternalError(
+            'Unhandled state: {}'.format(state))
+
+
+def _update_players(player_1, player_2, signer):
+    '''
+    Return: upd_player_1, upd_player_2
+    '''
+    if player_1 == '':
+        return signer, player_2
+
+    elif player_2 == '':
+        return player_1, signer
+
+    else:
+        return player_1, player_2
+
+
+def _update_board(board, space, mark):
+    index = space - 1
+
+    if board[index] != EMPTY_SPACE:
+        raise InvalidTransaction(
+            'Invalid Action: space {} already taken'.format(space))
+
+    return ''.join([curr if sqr != index else mark
+                    for sqr, curr in enumerate(board)])
+
+
+def _check_turn_and_get_mark(state, signer, player_1, player_2):
+    '''
+    Return: MARK_1 or MARK_2
+    '''
+    if state == P1_NEXT and signer == player_1:
+        return MARK_1
+    elif state == P2_NEXT and signer == player_2:
+        return MARK_2
+    else:
+        raise InvalidTransaction(
+            "Not this player's turn: {}".format(signer[:6]))
+
+
+def _get_rows(board):
+    return [
+        board[i:i + SIDE_LENGTH]
+        for i in range(0, SPACE_COUNT, SIDE_LENGTH)
+    ]
 
-        # Update game state
-        if _is_win(board, P1_MARK):
-            state = P1_WIN
-        elif _is_win(board, P2_MARK):
-            state = P2_WIN
-        elif EMPTY_SPACE not in board:
-            state = TIE
 
-        return board, state, player_1, player_2
+def _get_columns(board_rows):
+    return [
+        ''.join([row[i] for row in board_rows])
+        for i in range(SIDE_LENGTH)
+    ]
 
 
-def _is_win(board, letter):
-    wins = ((1, 2, 3), (4, 5, 6), (7, 8, 9),
-            (1, 4, 7), (2, 5, 8), (3, 6, 9),
-            (1, 5, 9), (3, 5, 7))
+def _get_diagonals(board_rows):
+    diag_left_above = ''.join([
+        board_rows[i][i]
+        for i in range(SIDE_LENGTH)
+    ])
 
-    for win in wins:
-        if (board[win[0] - 1] == letter
-                and board[win[1] - 1] == letter
-                and board[win[2] - 1] == letter):
-            return True
-    return False
+    diag_right_above = ''.join([
+        board_rows[SIDE_LENGTH - i - 1][i]
+        for i in range(SIDE_LENGTH)
+    ])
+
+    return diag_left_above, diag_right_above
+
+
+def _split_up_board(board):
+    rows = _get_rows(board)
+    columns = _get_columns(rows)
+    diagonals = _get_diagonals(rows)
+
+    return rows, columns, diagonals
+
+
+def _is_win(board, mark):
+    '''
```

Comment:
To me, the use of having a simple example like xo is to show how a transaction processor handles transactions and state data. The basic model that both xo and intkey follow is: get the transaction data, get the state data, do something, put the result back into state. From the perspective of showing off the system, it's relatively uninteresting what happens in the "do something" step, and the current documentation says so. It might even be more interesting if the implementations do differ, so long as they all produce the same output.